tests/topotests/pim_embedded_rp/test_pim_embedded_rp.py

```diff
@@ -0,0 +1,264 @@
+#!/usr/bin/env python3
+# SPDX-License-Identifier: ISC
+
+# Copyright (c) 2024 Network Education Foundation, Inc. ("NetDEF")
+#                    Rafael Zalamena
+
+import os
+import sys
+import pytest
+from functools import partial
+
+from lib import topotest
+
+# pylint: disable=C0413
+# Import topogen and topotest helpers
+from lib.topogen import Topogen, get_topogen
+from lib.topolog import logger
+
+from lib.pim import McastTesterHelper
+
+"""
+pim_embedded_rp.py: Test PIM embedded RP functionality.
+"""
+
+# Save the Current Working Directory to find configuration files.
+CWD = os.path.dirname(os.path.realpath(__file__))
+sys.path.append(os.path.join(CWD, "../"))
+
+# Required to instantiate the topology builder class.
+pytestmark = [pytest.mark.pim6d]
+
+
+def build_topo(tgen):
+    tgen.add_router("r1")
+    tgen.add_router("r2")
+    tgen.add_router("r3")
+
+    switch = tgen.add_switch("s1")
+    switch.add_link(tgen.gears["r1"])
+    switch.add_link(tgen.gears["r2"])
+
+    switch = tgen.add_switch("s2")
+    switch.add_link(tgen.gears["r1"])
+    switch.add_link(tgen.gears["r3"])
+
+    switch = tgen.add_switch("s3")
+    switch.add_link(tgen.gears["r2"])
+    switch.add_link(tgen.gears["r3"])
+
+    tgen.add_host("h1", "2001:DB8:100::100", "via 2001:DB8:100::1")
+    tgen.add_host("h2", "2001:DB8:200::100", "via 2001:DB8:200::1")
+
+    switch = tgen.add_switch("s10")
+    switch.add_link(tgen.gears["r2"])
+    switch.add_link(tgen.gears["h1"])
+
+    switch = tgen.add_switch("s20")
+    switch.add_link(tgen.gears["r3"])
+    switch.add_link(tgen.gears["h2"])
+
+
+app_helper = McastTesterHelper()
+
+
+def setup_module(mod):
+    tgen = Topogen(build_topo, mod.__name__)
+    tgen.start_topology()
+
+    router_list = tgen.routers()
+    for _, router in router_list.items():
+        router.load_frr_config(os.path.join(CWD, f"{router.name}/frr.conf"))
+
+    tgen.start_router()
+    app_helper.init(tgen)
+
+
+def teardown_module(mod):
+    tgen = get_topogen()
+    tgen.stop_topology()
+
+
+def test_ospfv3_convergence():
+    "Wait for OSPFv3 protocol convergence"
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+
+    logger.info("waiting for protocols to converge")
```

Comment:
FYI: `step()` is a good alternative for marking/logging important events in a test.